### PR TITLE
Cleanup, optimization, and bug fixes

### PR DESCRIPTION
**Pull request by toofishes**
The main big points here:
- Use mutexes correctly so we don't have to use volatile
- Use the 'static' keyword where appropriate so the compiler can inline and optimize as it sees fit
- Fix a rather large issue in the playback rate and drift calculation

This also includes the one commit from Stef's outstanding pull request.

https://github.com/albertz/shairport/pull/131

### CODE REVIEW

**Comment by abrasive** [comment]
Thanks for all the cleanups!

I don't see how using a mutex (for eg. the volume control) has any bearing on marking things volatile.

Certainly I'm a little over-eager with volatile (courtesy of a lot of pain with embedded compilers), but I wasn't aware of any mutex semantics that cause the compiler to reconsider its variable values!

Also, there's no point to using a mutex with the volume control at all - it doesn't matter whether it changes mid-block or not.

**Comment by toofishes** [comment]
Calling pthread_mutex_lock() and unlock() guarantee memory synchronization: http://pubs.opengroup.org/onlinepubs/9699919799/basedefs/V1_chap04.html#tag_04_11

Right now, you are doing something absolutely silly- reading from the volatile variable 'volume' in a tight loop, which requires a memory access every single time, as the compiler is held hostage and can't just store that value in a register.

volatile is like using a bomb when you need a flyswatter; it has some very bad consequences, and I'm still not sure it actually gives you the guarantees you are looking for.

So in short, the mutex is much much cheaper than the volatile in all respects.
- http://stackoverflow.com/questions/78172/using-c-pthreads-do-shared-variables-need-to-be-volatile
- http://stackoverflow.com/questions/3208060/does-guarding-a-variable-with-a-pthread-mutex-guarantee-its-also-not-cached

**Comment by abrasive** [comment]
Interesting. Thank you.
